libraries/AP_Proximity/AP_Proximity_Backend.cpp

```diff
@@ -351,40 +330,32 @@ bool AP_Proximity_Backend::get_next_ignore_start_or_end(uint8_t start_or_end, in
     return found;
 }
 
-void AP_Proximity_Backend::update_locations()
+// returns true if database is ready to be pushed to and all cached data is ready
+bool AP_Proximity_Backend::database_prepare_for_push()
 {
-    WITH_SEMAPHORE(_rsem);
+    if (!AP::ahrs().get_position(current_loc)) {
+        return false;
+    }
 
-    // get number of objects, return early if none
-    const uint8_t num_objects = get_object_count();
-    if (num_objects == 0) {
-        _location_count = 0;
+    current_vehicle_bearing = AP::ahrs().yaw_sensor * 0.01f;
+    return true;
+}
+
+// pre-req: database_prepare_for_push() has just returned true in same thread/function
+// update Object Avoidance database with Earth-frame point
+void AP_Proximity_Backend::database_push(const float angle, const float distance, const uint32_t timestamp_ms, const bool needs_prepare)
+{
+    AP_OAPathPlanner *oa = AP_OAPathPlanner::get_singleton();
+    if (oa == nullptr || !oa->database_enabled()) {
         return;
     }
-
-    // get vehicle location and heading in degrees
-    float veh_bearing;
-    Location my_loc;
-    {
-        WITH_SEMAPHORE(AP::ahrs().get_semaphore());
-        if (!AP::ahrs().get_position(my_loc)) {
-            return;
-        }
-        veh_bearing = AP::ahrs().yaw_sensor * 0.01f;
+    if (needs_prepare && !database_prepare_for_push()) {
+        return;
     }
 
-    // convert offset from vehicle position to earth frame location
-    const uint32_t now_ms = AP_HAL::millis();
-    _location_count = 0;
-    for (uint8_t i=0; i<num_objects; i++) {
-        float ang_deg, dist_m;
-        if (get_object_angle_and_distance(i, ang_deg, dist_m)) {
-            Location temp_loc = my_loc;
-            temp_loc.offset_bearing(wrap_180(veh_bearing + ang_deg), dist_m);
-            _locations[_location_count].loc = temp_loc;
-            _locations[_location_count].radius_m = 1;
-            _locations[_location_count].last_update_ms = now_ms;
-            _location_count++;
-        }
-    }
+    // current_loc and current_vehicle_bearing have
+    // been cached from database_prepare_for_push()
+    Location temp_loc = current_loc;
```

Comment:
If prepare got brought inside this function then the cached current_loc can be removed which saves ram/copying and clearer.

Reply:
Most of the Proxy drivers do this, but there's a MAVLink msg that would run this 72 times so we don't need to recacl the heading 72 times.